packages/periphery/contracts/oracles/UniswapV2Oracle.sol

```diff
@@ -0,0 +1,182 @@
+
+    uint256 _granularity = granularity;
+    address _factory = factory;
+    uint256 _windowSize = windowSize;
+    for (uint i; i < length; ) {
+      PairOracleOptions memory options = newOptions[i];
+      if (options.secondsAgo < _granularity || options.secondsAgoLiquidation < _granularity) revert WrongValue();
+      if (options.secondsAgo > _windowSize || options.secondsAgoLiquidation > _windowSize) revert WrongValue();
+
+      TokenPair memory tokenPair = tokenPairs[i];
+      address pair = IUniswapV2Factory(_factory).getPair(tokenPair.token0, tokenPair.token1);
+      if (pair == address(0)) revert PairNotFound();
+      if (pairOptions[pair].secondsAgo != 0) revert PairAlreadyExists();
+
+      pairs.push(pair);
+      pairOptions[pair] = options;
+
+      // populate the array with empty observations (first call only)
+      for (uint j; j < _granularity; ) {
+        pairObservations[pair].push();
+
+        unchecked {
+          j++;
+        }
+      }
+
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  /// @notice Remove a pair from the oracle
+  function removePairAt(uint256 index) external onlyOwner {
+    uint256 lastIndex = pairs.length - 1;
+    address pair = pairs[index];
+    delete pairOptions[pair];
+    delete pairObservations[pair];
+
+    if (lastIndex != 0 && lastIndex != index) {
+      pairs[index] = pairs[lastIndex];
+    }
+
+    pairs.pop();
+  }
+
+  /// @notice Returns the price in Q96 format of the given pair in the last `secondsAgo` seconds
+  function computePrice(address pair, uint16 secondsAgo) public view returns (uint256) {
+    uint256 observationIndex = observationIndexOf(block.timestamp - secondsAgo);
+    Observation memory observationStart = pairObservations[pair][observationIndex];
+
+    uint timeElapsed = block.timestamp - observationStart.timestamp;
+    if (timeElapsed > windowSize) revert MissingHistoricalObservation();
+    // should never happen.
+    if (timeElapsed < granularity) revert UnexpectedTimeElapsed();
+    (uint price0CumulativeEnd, , ) = UniswapV2OracleLibrary.currentCumulativePrices(pair);
+
+    // ">>16" converting from q112 to q96 FP
+    return uint256((price0CumulativeEnd - observationStart.price0Cumulative) / timeElapsed) >> 16;
+  }
+
+  function getBalancePrice(address quoteToken, address baseToken) external view returns (uint256 price) {
+    address pair = IUniswapV2Factory(factory).getPair(quoteToken, baseToken);
+
+    price = computePrice(pair, pairOptions[pair].secondsAgo);
+    if (quoteToken < baseToken) {
+      price = Math.mulDiv(X96ONE, X96ONE, price);
+    }
+  }
+
+  function getMargincallPrice(address quoteToken, address baseToken) external view returns (uint256 price) {
+    address pair = IUniswapV2Factory(factory).getPair(quoteToken, baseToken);
+
+    price = computePrice(pair, pairOptions[pair].secondsAgoLiquidation);
+    if (quoteToken < baseToken) {
+      price = Math.mulDiv(X96ONE, X96ONE, price);
```

Comment:
This is an estimation, which will be different from the precise calculation (using price1Cumulative)